File: wk8/java/inverseFactorial.java
// problem link: https://open.kattis.com/problems/inversefactorial

import java.io.*;
import java.math.*;
import java.util.*;

public class inversefactorial {
    public static void main(String[] args) throws IOException {
        BufferedReader r = new BufferedReader(new InputStreamReader(System.in));
        String n = r.readLine();
		if(n.length() < 4) {
		    int nn = Integer.parseInt(n), i = 1, c = 1;
		    while(c != nn)
		        c *= ++i;
		    System.out.println(i);
		    return;
		}
		int i = 1;
		double d = 1.0;
		while(Math.floor(d) < n.length())
			d += Math.log10(i++);
		System.out.println(i - 1);
    }
}

File: wk8/cpp/inverseFactorial.cpp
// problem link: https://open.kattis.com/problems/inversefactorial

#include <bits/stdc++.h>
using namespace std;

int main()
{
    ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    
    string n;
    cin >> n;
    if(n.size() < 4) {
        int nn = stoi(n), i = 1, c = 1;
        while(c != nn)
            c *= ++i;
        cout << i;
        return 0;
    }
    int i = 1;
    double d = 1.0;
    while(floor(d) < n.size())
        d += log10(i++);
    cout << i - 1;

    return 0;
}